File: web/apps/dashboard/lib/trpc/routers/logdrain/delete.ts
import { insertAuditLogs } from "@/lib/audit";
import { and, db, eq, schema } from "@/lib/db";
import { TRPCError } from "@trpc/server";
import { z } from "zod";
import { workspaceProcedure } from "../../trpc";

export const deleteLogdrain = workspaceProcedure
  .input(z.object({ id: z.string().min(1) }))
  .mutation(async ({ ctx, input }) => {
    try {
      await db.transaction(async (tx) => {
        const [drain] = await tx
          .select({ id: schema.logdrains.id, name: schema.logdrains.name })
          .from(schema.logdrains)
          .where(
            and(
              eq(schema.logdrains.id, input.id),
              eq(schema.logdrains.workspaceId, ctx.workspace.id),
            ),
          )
          .for("update");
        if (!drain) {
          throw new TRPCError({ code: "NOT_FOUND", message: "Log drain not found" });
        }

        await tx.delete(schema.logdrainState).where(eq(schema.logdrainState.logdrainId, drain.id));
        await tx
          .delete(schema.logdrains)
          .where(
            and(
              eq(schema.logdrains.id, input.id),
              eq(schema.logdrains.workspaceId, ctx.workspace.id),
            ),
          );
        await insertAuditLogs(tx, {
          workspaceId: ctx.workspace.id,
          actor: { type: "user", id: ctx.user.id },
          event: "logdrain.delete",
          description: `Deleted log drain ${input.id}`,
          resources: [],
          context: { location: ctx.audit.location, userAgent: ctx.audit.userAgent },
        });
      });
      return { id: input.id };
    } catch (error) {
      if (error instanceof TRPCError) {
        throw error;
      }
      console.error("Failed to delete log drain", error);
      throw new TRPCError({ code: "INTERNAL_SERVER_ERROR", message: "Failed to delete log drain" });
    }
  });

File: svc/api/routes/v2_logdrains/delete.go
package logdrains

import (
	"context"
	"net/http"

	"github.com/unkeyed/unkey/internal/services/auditlogs"
	"github.com/unkeyed/unkey/pkg/auditlog"
	"github.com/unkeyed/unkey/pkg/codes"
	"github.com/unkeyed/unkey/pkg/db"
	"github.com/unkeyed/unkey/pkg/fault"
	"github.com/unkeyed/unkey/pkg/rbac"
	"github.com/unkeyed/unkey/pkg/rbac/permissions"
	"github.com/unkeyed/unkey/pkg/urn"
	"github.com/unkeyed/unkey/pkg/zen"
	"github.com/unkeyed/unkey/svc/api/openapi"
)

type Delete struct {
	DB        db.Database
	Auditlogs auditlogs.AuditLogService
}

func (h *Delete) Method() string { return http.MethodPost }
func (h *Delete) Path() string   { return "/v2/logdrains.deleteLogdrain" }
func (h *Delete) Handle(ctx context.Context, s *zen.Session) error {
	principal, err := s.GetPrincipal()
	if err != nil {
		return err
	}
	req, err := zen.BindBody[openapi.LogdrainIdRequest](s)
	if err != nil {
		return err
	}
	if err := principal.Authorize(rbac.U(urn.V1{WorkspaceID: principal.AuthorizedWorkspaceID, Resource: "logdrains/" + req.LogdrainId}, permissions.Delete)); err != nil {
		return err
	}
	err = db.TxRetry(ctx, h.DB.RW(), func(ctx context.Context, tx db.DBTX) error {
		row, err := db.Query.LockLogdrain(ctx, tx, db.LockLogdrainParams{WorkspaceID: principal.AuthorizedWorkspaceID, ID: req.LogdrainId})
		if err != nil {
			if db.IsNotFound(err) {
				return fault.Wrap(err, fault.Code(codes.Data.Logdrain.NotFound.URN()), fault.Public("Log drain not found."))
			}
			return err
		}
		if err := db.Query.DeleteLogdrain(ctx, tx, db.DeleteLogdrainParams{WorkspaceID: principal.AuthorizedWorkspaceID, ID: row.ID}); err != nil {
			return err
		}
		return h.Auditlogs.Insert(ctx, tx, []auditlog.AuditLog{{
			WorkspaceID: principal.AuthorizedWorkspaceID, Event: "logdrain.delete", Display: "Deleted log drain " + row.ID,
			ActorID: principal.Subject.ID, ActorType: auditlog.AuditLogActor(principal.Subject.Type), ActorName: principal.Subject.Name,
			ActorMeta: map[string]any{}, CorrelationID: "", RemoteIP: s.Location(), UserAgent: s.UserAgent(),
			Resources: []auditlog.AuditLogResource{{ID: row.ID, Type: "logdrain", Name: row.Name, DisplayName: "", Meta: map[string]any{}}},
		}})
	})
	if err != nil {
		return err
	}
	var response openapi.LogdrainMutationResponse
	response.Meta.RequestId = s.RequestID()
	response.Data.Id = req.LogdrainId
	return s.JSON(http.StatusOK, response)
}
